src/cuckoocache.h

```diff
@@ -0,0 +1,482 @@
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_CUCKOOCACHE_H
+#define BITCOIN_CUCKOOCACHE_H
+
+#include <array>
+#include <atomic>
+#include <cmath>
+#include <cstring>
+#include <memory>
+#include <utility>
+#include <vector>
+
+
+/** High-performance cache primitives.
+ *
+ * Summary:
+ *
+ * 1. @ref bit_packed_atomic_flags is bit-packed atomic flags for garbage collection
+ *
+ * 2. @ref cache is a cache which is performant in memory usage and lookup speed. It
+ * is lockfree for erase operations. Elements are lazily erased on the next insert.
+ */
+namespace CuckooCache
+{
+/** @ref bit_packed_atomic_flags implements a container for garbage collection flags
+ * that is only thread unsafe on calls to setup. This class bit-packs collection
+ * flags for memory efficiency.
+ *
+ * All operations are `std::memory_order_relaxed` so external mechanisms must
+ * ensure that writes and reads are properly synchronized.
+ *
+ * On setup(n), all bits up to `n` are marked as collected.
+ *
+ * Under the hood, because it is an 8-bit type, it makes sense to use a multiple
+ * of 8 for setup, but it will be safe if that is not the case as well.
+ */
+class bit_packed_atomic_flags
+{
+    std::unique_ptr<std::atomic<uint8_t>[]> mem;
+
+public:
+    /** No default constructor, as there must be some size. */
+    bit_packed_atomic_flags() = delete;
+
+    /**
+     * bit_packed_atomic_flags constructor creates memory to sufficiently
+     * keep track of garbage collection information for `size` entries.
+     *
+     * @param size the number of elements to allocate space for
+     *
+     * @post bit_set, bit_unset, and bit_is_set function properly forall x. x <
+     * size
+     * @post All calls to bit_is_set (without subsequent bit_unset) will return
+     * true.
+     */
+    bit_packed_atomic_flags(uint32_t size)
+    {
+        // pad out the size if needed
+        size = (size + 7) / 8;
+        mem.reset(new std::atomic<uint8_t>[size]);
+        for (uint32_t i = 0; i < size; ++i)
+            mem[i].store(0xFF);
+    };
+
+    /** setup marks all entries and ensures that bit_packed_atomic_flags can store
+     * at least `b` entries.
+     *
+     * @param b the number of elements to allocate space for
+     * @post bit_set, bit_unset, and bit_is_set function properly forall x. x <
+     * b
+     * @post All calls to bit_is_set (without subsequent bit_unset) will return
+     * true.
+     */
+    inline void setup(uint32_t b)
+    {
+        bit_packed_atomic_flags d(b);
+        std::swap(mem, d.mem);
+    }
+
+    /** bit_set sets an entry as discardable.
+     *
+     * @param s the index of the entry to bit_set
+     * @post immediately subsequent call (assuming proper external memory
+     * ordering) to bit_is_set(s) == true.
+     */
+    inline void bit_set(uint32_t s)
+    {
+        mem[s >> 3].fetch_or(1 << (s & 7), std::memory_order_relaxed);
```

Comment:
I am skeptical that using `std::memory_order_relaxed` here is sufficient for the usage in the cuckoo cache implementation. Most *correct* lock-free algorithms I've seen use release/acquire memory ordering. Note that if it weren't sufficient, then the resulting problems might not show up on x86, but could show up on architectures with a weaker native memory ordering model.

Reply:
Comments about this on the upstream PR indicate that it was an intentional choice to use relaxed here: https://github.com/bitcoin/bitcoin/pull/8895#discussion_r86878499

Reply:
Also documented as such in the help text:
```cpp
 * All operations are `std::memory_order_relaxed` so external mechanisms must
 * ensure that writes and reads are properly synchronized.
```